rfcs/text/0005_route_handler.md

```diff
@@ -0,0 +1,187 @@
+```
+
+##### Redirection
+The server wants a user to perform additional actions.
+```typescript
+const kibanaResponseToolkit = {
+  redirected: (url: string, options?: HttpResponseOptions) => new KibanaResponse(302, url, options),
+  notModified: (options?: HttpResponseOptions) => new KibanaResponse(304, undefined, options),
+```
+
+#####  Error
+Server signals that request cannot be handled and explains details of the error situation
+```typescript
+// Supports attaching additional data to send to the client
+interface ResponseError extends Error {
+  meta?: {
+    data?: JSONValue;
+    errorCode?: string; // error code to simplify search, translations in i18n, etc.
+    docLink?: string; // link to the docs
+  }
+}
+
+export const createResponseError = (error: Error | string, meta?: ResponseErrorType['meta']) =>
+  new ResponseError(error, meta)
+
+const kibanaResponseToolkit = {
+  // Client errors
+  badRequest: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(400, err, options),
+  unauthorized: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(401, err, options),
+
+  forbidden: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(403, err, options),
+  notFound: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(404, err, options),
+  conflict: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(409, err, options),
+
+  // Server errors
+  internal: <T extends ResponseError>(err: T, options?: HttpResponseOptions) =>
+    new KibanaResponse(500, err, options),
+```
+
+##### Custom
+If a custom response is required
+```typescript
+interface CustomOptions extends HttpResponseOptions {
+  statusCode: StatusCode;
+}
+export const kibanaResponseToolkit = {
+  custom: <T extends HttpResponsePayload>(payload: T, {statusCode, ...options}: CustomOptions) =>
+    new KibanaResponse(statusCode, payload, options),
+```
+# Drawbacks
+- `Handler` is not compatible with Legacy platform implementation when anything can be returned or thrown from handler function and server send it as a valid result. Transition to the new format may require additional work in plugins.
+- `Handler` doesn't cover **all** functionality of the Legacy server at the current moment. For example, we cannot render a view in New platform yet and in this case, we have to proxy the request to the Legacy platform endpoint to perform rendering. All such cases should be considered in an individual order.
+- `KibanaResponseToolkit` may not cover all use cases and requires an extension for specific use-cases.
+- `KibanaResponseToolkit` operates low-level Http primitives, such as Headers e.g., and it is not always handy to work with them directly.
+- `KibanaResponse` cannot be extended with arbitrary data. 
```

Comment:
It's not a drawback per se. 
We have `custom` helper as escape-hatch, but if one doesn't cover a very weird use case it will take some time to extend the core.